Add Services V2

brb opened the pull request:
This PR introduces [Services V2](https://docs.google.com/document/d/1zAqme6Blk-brIG6K-Y_1k3lpUB2Ql0xXwpvR0YpCLNs/edit).

TL;DR To avoid the service churn, a new BPF map (`lb{4,6}_backends`) to store service endpoints is introduced. This allows us to directly point to a service endpoint address in a CT entry without the need to lookup a service map first. So, the service map entries order can change without disrupting any existing connections. To support non-disruptive upgrade/downgrade of the cilium-agent, we keep and maintain both types of service maps at the same time, and we point to both type entries in CT.

NOTES:

* This PR can be reviewed per-commit basis. Please read commit messages as well. If you are not happy with the number of changes or commit structuring, I can give another try or/and create a few smaller PRs.

* There is quite a bit of copy-paste. Mainly, because I didn't want to extend the existing service interfaces which will be removed (with a lot of code of this PR) in the 1.6.

* Although I've introduced the L4 protocol field to the svc v2, this is currently not used. I didn't want to make this change even more complex. We can support different L4 protocols for svc in the future.

* For backend ID type I'm using `u16` which sets a limit of max 65535 backends (service endpoints) per cluster. It's not too late, I can easily switch to `u32` to support more backends. UPDATE: will change to `u32` in a subsequent PR.

* Currently, we maintain both types of services at the cost of larger memory footprint and more expensive service lookups. I'm adding the `--disable-legacy-services` to the cilium-agent to disable legacy services (will come in a separete PR). Also, in that PR I'm going to change `cilium bpf lb list` to use svc v2 instead of legacy as a source. UPDATE: flag will be named `--enable-legacy-services` with a default set to true.

* I've been testing the upgrade/downgrade only manually. I will add an integration test to test it (in separate PR).

* Currently, a svc update can change the order of svc v2 endpoints. This does not break anything, but probably it's better if a packet with the same hash will go to the same endpoint as in a previous connection, assuming that no deletion of the svc endpoints has happened (will do in separate PR).

* There are many comments by the Hound bot about missing comments for public types / functions or methods. For some it feels a bit stupid to add. If you think otherwise, I'm happy to add the missing dummy comments.

Fix #7106.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7540)
<!-- Reviewable:end -->

brb committed:
bpf,lbmap: Add underlying types and BPF maps for service v2

Service v2 (or svc v2) is a new type of services which aims to resolve
the following shortcomings of the current type of services (we refer
to the latter as "legacy"):

1. Duplicate service endpoint entries due to the implementation limitations.
2. No support for L4 protocol.

The svc v2 solves the first limitation by introducing a new table for
storing unique set of service endpoints (also known as backends). This
allows a connection tracking table entries to refer to service endpoints
directly, i.e. without consulting a corresponding entry in service table
first.

The second limitation is addressed by introducing the protocol field
in the service entry key.

The design document can be found at: https://docs.google.com/document/d/1zAqme6Blk-brIG6K-Y_1k3lpUB2Ql0xXwpvR0YpCLNs/edit

This commit introduces the underlying types for the svc v2 w/o adding
a logic to the agent/datapath code to support them.

The `ServiceKeyV2` and `ServiceValueV2` related types are copy-pasted from
`ServiceKey/ServiceValue` with small modifications applied. An alternative
approach would have been to extend the latter types. But, considering
that legacy services won't be supported in the 1.6 release, it makes
easier to just copy-paste them.

Next, this commit introduces a new data type called `BackendAddrID`
which is used to uniquely identify a service endpoint when its real ID
(aka backend ID) is not known. `BackendAddrID` is in the form of
"IP:PORT".

Finally, this commit creates the required maps for the svc v2. Keep in
mind, that at the same time we have to support both types of services
(due to possible downgrade/upgrade of cilium-agent), thus two instances
of the maps.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
service: Add methods to acquire backend ID for svc v2

This commit does two things:

1. Restructure the local ID acquisition code in a way that it can
   be reused by other entities to acquire node-local IDs.
2. Add methods to acquire node-local backend IDs for svc v2.

The backend IDs are going to be used in BPF maps to uniquely
identify service v2 endpoints. So, they cannot be of the
`BackendAddrID` type. Therefore, we reuse the mechanism used
by `ServiceID` to generate them. Luckily, the backend ID's
scope is node-local, so the ID generation does not have to
rely on kvstore which would make their acquisition much more
slow.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
lbmap: Move interface types to types.go

To reduce lbmap.go module size to increase readability.

No functional changes.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
lbmap: Extend Service{4,6}Value types

Add `GetPort`, `IsIPv6` and `BackendAddrID` methods which are
going to be used for migrating services from legacy to v2.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
lbmap: Move type conversion routines to types.go

Again, to keep the module size sane.

No functional changes.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
lbmap: Add type conversion routines for svc v2 types

The routines are not used yet.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
loadbalancer: Add BackendID

`BackendID` is used to uniquely identify a service endpoint which
can be used by many services.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
logfields: Add logfields for svc v2 logging

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
lbmap: Use BackendAddrID in the bpf_service cache

For safety reason, let's use the BackendAddrID.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
lbmap: Add backend ref-counting to the cache

We need to reference count backend instances, as the same
entry in the backend BPF map can be referred by many services.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
lbmap: Extend per-service cache

Add two new fields to the per-service cache:

- `slaveSlotByBackendAddrID` - used to map backend => backend position
  in the corresponding legacy svc map.
- `backendV2` - to cache svc v2 entries.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
lbmap: Extend global service cache

Add the `backendIDByAddrID` field to map backend addr ID => backend ID.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
lbmap: Create svc v2 entries with UpdateService

This commit makes the `UpdateService` function to create service entries
for two types (legacy and v2) at the same time.

This requires the following changes:

- Make `prepareUpdate` to return backend IDs which needs to be
  acquired/added and released/removed.
- Set `count` field in the legacy svc entry to point to the backend ID
  (the backward-compatibility).
- Add entries in the v2 BPF maps (services and backends). The service
  entry's `count` field points to the slave number in a corresponding
  service.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon,lbmap: Restore and migrate v2 services

This commit makes restoration and migration of v2 services possible
by doing the following:

- Introducing dumping routines of svcv2-related maps. Also, one
  noticeable change that any dump routine does not include
  master service entry in the returned maps.
- Adding a helper function to restore backend IDs before doing
  the restoration of the service entries.
- Creating v2 from legacy.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon,lbmap: Remove svc v2 when deleting legacy svc

The main difference between removing the legacy svc is
that we might need to remove relevant backend entries
depending whether they are referenced by any other
service. Also, when removing the backend, we need to
release its ID.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
bpf: Add support for svc v2 to datapath

This commit introduces svc v2 support to the BPF datapath (to both -
lb and lxc).

First, this makes the svc lookup to rely only on services of
the v2 type. This is safe, as before reloading the datapath,
all v2 services will be created from legacy ones by the daemon.

Second, as we need to support upgrading/downgrading of cilium-agent
without disrupting open connections, both legacy and v2 service
endpoints are stored in CT entries.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

joestringer commented on common/types/ipv4.go:
```diff
@@ -25,6 +25,12 @@ func (v4 IPv4) IP() net.IP {
 	return v4[:]
 }
 
+func (v4 IPv4) DuplicateIP() net.IP {
```
I can't figure out how this is actually used (it's invoked, but I can't tell that it makes a meaningful difference). Seems to me that `types.IPv4` is defined as a golang array (which I would assume is by-value), and we don't use pointers to addresses, we just copy them around. Is this necessary? (If so, might be worth searching for `copy` in this PR for more places to use this..)

brb replied:
>  <..> which I would assume is by-value

You are right. Just checked that modifying a returned IP addr by `types.IPv4.IP()` does not modify the original IP addr.

I've removed the methods and their usage.

joestringer commented on bpf/lib/common.h:
```diff
@@ -476,6 +476,35 @@ struct lb6_service {
+	__be16 dport;		/* L4 port filter, if unset, all ports apply */
+	__u16 slave;		/* Backend iterator, 0 indicates the master service */
+	__u8 proto;		/* L4 protocol, currently not used (set to 0) */
+	__u8 pad;
```
This seems to be aligning to a 2B boundary, total size 22B per my `pahole`. Were you intending to align this to a 64-bit boundary? (Similarly for the `lb4_key_v2`)

brb replied:
My bad - I've been using the `packed` attribute (removed), and I've never checked what layout the compiler generates w/o the attribute being set.

Anyway, after the change we have the following layouts:
```
struct lb4_key_v2 {
        uint32_t                   address;              /*     0     4 */
        uint16_t                   dport;                /*     4     2 */
        uint16_t                   slave;                /*     6     2 */
        uint8_t                    proto;                /*     8     1 */
        uint8_t                    pad[3];               /*     9     3 */

        /* size: 12, cachelines: 1, members: 5 */
        /* last cacheline: 12 bytes */
};
struct lb6_key_v2 {
        union v6addr               address;              /*     0    16 */
        uint16_t                   dport;                /*    16     2 */
        uint16_t                   slave;                /*    18     2 */
        uint8_t                    proto;                /*    20     1 */
        uint8_t                    pad[3];               /*    21     3 */

        /* size: 24, cachelines: 1, members: 5 */
        /* last cacheline: 24 bytes */
};
```

BTW, is there any requirement for alignment (AFAIK, x86_64 doesn't enforce one)?

joestringer replied:
I *suspect* this was previously more about our attempts to match alignment between the C side here and Golang structures, rather than any other alignment requirement. (/cc @borkmann )

joestringer commented on bpf/lib/conntrack.h:
```diff
@@ -225,6 +225,16 @@ static inline int __inline__ __ct_lookup(void *map, struct __sk_buff *skb,
+			 * number is stored within the "ct_entry.slave" field, while
+			 * for the v2, a backend id is stored in the "ct_entry.rx_bytes"
+			 * field which is not used by a entry with dir=CT_SERVICE.
+			 */
```
nit: Can we make this official by adjusting the `ct_entry` structure with a union and put the comment there? (Be ware, the verifier might actually complain about this, I don't mind if this is a separate follow-up)

Also, on the CLI side it would be useful to format this properly for CT service entries.

brb replied:
Makes sense, adding this to my TODO to address in a separate PR.

ianvernon commented on pkg/maps/lbmap/ipv4.go:
```diff
@@ -40,6 +41,36 @@ var (
+				return nil, nil, err
+			}
+
+			return svcKey.ToNetwork(), svcVal.ToNetwork(), nil
```
Shouldn't the dump function convert it to the `ToHost` representation?

brb replied:
Good spot and good question. I was (blindly) following the existing code. `ToNetwork()` it will just do the swap without checking whether it has been swapped before, so it has the same effect as `ToHost()`. I'm adding to my TODO to fix this in separate PR.

joestringer commented on bpf/lib/conntrack.h:
```diff
@@ -682,8 +760,19 @@ static inline int __inline__ ct_create4(void *map, struct ipv4_ct_tuple *tuple,
+	 * Previously, the rx_bytes field was not used for entries with
+	 * the dir=CT_SERVICE (see GH#7060). Therefore, we can safely abuse
+	 * this field to save the backend_id. The hack will go away once we stop
+	 * supporting the legacy svc.
```
Can you explain the plan of attack for this?

Is it something like this?
* v1.5 populates the `slave` with v1 backend, `rx_bytes` with v2
* in v1.6, we populate both `rx_bytes` and `slave` with the v2 backend,
* In v1.7 we can finally remove the `rx_bytes` hack

brb replied:
Exactly this is the plan. Updated the comment.

joestringer commented on pkg/service/id.go:
```diff
@@ -61,21 +63,60 @@ func DeleteID(id uint32) error {
+	}
+
+	// TODO(brb) This shouldn't happen (otherwise, there is a bug in the code).
+	//           But maybe it makes sense to delete all svc v2 in this case.
```
Is there something left to do here?

brb replied:
This TODO is still relevant, as I haven't implemented the removal of svc when the backend ID acquisition fails. Error handling for svc v2 needs a better thought though.

joestringer commented on pkg/maps/lbmap/types.go:
```diff
@@ -0,0 +1,522 @@
+		beIP = svc6Val.Address.IP()
+		bePort = svc6Val.Port
+		beWeight = svc6Val.Weight
+		beID = svc6Val.Count
```
The commit which moves this function to this file states "No functional changes", however this beID field appears to be new. Is there any potential impact of that?

brb replied:
Ouch, this slipped unnoticed. Good catch. Actually setting the `beID` is not needed, so I removed it.

ianvernon commented on pkg/maps/lbmap/types.go:
```diff
@@ -426,3 +426,97 @@ func revNatValue2L3n4AddrID(revNATKey RevNatKey, revNATValue RevNatValue) *loadb
 }
+
+// LBSVC2ServiceKeynValuenBackendValueV2 transforms the SVC Cilium type into a bpf SVC v2 type.
+func LBSVC2ServiceKeynValuenBackendV2(svc *loadbalancer.LBSVC) (ServiceKeyV2, []ServiceValueV2, []Backend, error) {
```
While this carries over the convention from the old code, this function naming is really hard to read. Why not `ConvertLBSVC`? It's more concise, and what it is converted to is specified in the values that are returned.

brb replied:
Agreed, added to my TODO and will do in separate PR.

joestringer commented on pkg/maps/lbmap/types.go:
```diff
@@ -0,0 +1,522 @@
+	log.WithFields(logrus.Fields{
+		"lbFrontend": svc.FE.String(),
+		"lbBackend":  svc.BES,
+	}).Debug("converting Cilium load-balancer service (frontend -> backend(s)) into BPF service")
```
nit: Might be useful to mention `v2` in these logs so it's clear which conversion is occurring (similarly for all other v2 conversion functions here).

brb replied:
Sure, added.

ianvernon commented on pkg/maps/lbmap/types.go:
```diff
@@ -426,3 +426,97 @@ func revNatValue2L3n4AddrID(revNATKey RevNatKey, revNATValue RevNatValue) *loadb
+}
+
+// serviceKey2L3n4Addr converts the given svcKey to a L3n4Addr.
+func serviceKey2L3n4AddrV2(svcKey ServiceKeyV2) *loadbalancer.L3n4Addr {
```
Same here for naming. This can be handled outside of this PR to reduce the amount of churn, as this is nit-picky on my part :)

brb replied:
Same ^^.

joestringer commented on pkg/maps/lbmap/bpfservice.go:
```diff
@@ -195,29 +243,169 @@ func (l *lbmapCache) prepareUpdate(fe ServiceKey, backends []ServiceValue) *bpfS
+
+// addBackendV2Locked increments a ref count for the given backend and returns
+// whether any instance of the backend existed before.
+func (l *lbmapCache) addBackendV2Locked(addrID BackendAddrID) bool {
```
nit: It'd be nice if we could just compose this reference counting via a structure in `pkg/counter/` for code re-use, but that can be a "wishlist" item for future.

brb replied:
Thanks, I was not aware of the `pkg/counter`. Adding to my TODO.

joestringer commented on pkg/maps/lbmap/bpfservice.go:
```diff
@@ -44,36 +46,49 @@ type bpfService struct {
+	slaveSlotByBackendAddrID map[BackendAddrID]int
+
+	// backendsV2 is a map of all service v2 backends indexed by the legacy IDs.
+	// TODO(brb) use list instead to preserve the ordering when svc backends change.
```
What's the plan for this TODO?

brb replied:
It's coming in a separate PR, as mentioned in the NOTES of this PR:

```
Currently, a svc update can change the order of svc v2 endpoints. This does not break anything, but probably it's better if a packet with the same hash will go to the same endpoint as in a previous connection, assuming that no deletion of the svc endpoints has happened (will do in separate PR).
```

ianvernon commented on pkg/maps/lbmap/bpfservice.go:
```diff
@@ -277,3 +297,38 @@ func (l *lbmapCache) delBackendV2Locked(addrID BackendAddrID) (bool, error) {
+}
+
+func (l *lbmapCache) getBackendIDByAddrID(addrID BackendAddrID) uint16 {
+	l.mutex.Lock()
```
This could be an `RLock` if you want.

brb replied:
Sure, but I would need to replace `lbmapCache.mutex` with `lock.RWMutex`. I guess it's fine to replace?

joestringer replied:
Probably only worth doing if we expect readers to be far more common than writers?

brb replied:
Updated to use `lock.RWMutex`.

joestringer commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -348,15 +247,55 @@ func updateMasterService(fe ServiceKey, nbackends int, nonZeroWeights uint16) er
 	)
 
-	svc := cache.prepareUpdate(fe, backends)
+	// Acquire missing backend IDs
```
What's a missing backend ID?

brb replied:
I've updated the comment to:

```
// Acquire IDs for new backends (i.e. the ones which do not exist yet and will
// be created in this function)
```

joestringer commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -348,15 +247,55 @@ func updateMasterService(fe ServiceKey, nbackends int, nonZeroWeights uint16) er
+		addr := *serviceValue2L3n4Addr(newBackendsByAddrID[addrID])
+		backendID, err := acquireBackendID(addr)
+		if err != nil {
+			return fmt.Errorf("Unable to acquire backend ID for %s: %s", addrID, err)
```
Acquisition of some backend IDs may occur in this case, and there is no corresponding release.

brb replied:
I was thinking about adding a code to release backend IDs in the case of an error, but in the end I didn't add it. Mainly because the local ID allocator returns the same ID for a backend which previously acquired and didn't release it. So, e.g. if a svc add request fails, a retry request will not acquire new IDs for the same backends.

Also in theory, `UpdateService` should never fail unless there is a bug in the code.

joestringer replied:
Does this cause a resource management problem long-term? If you cycle backends enough times, then the ID allocator would eventually run out of IDs. It's keyed by IP address + port so difficult with IPv4 to run out of entries but plausible (10.x.y.z/8 = 24 bits + port = 16 bits + proto)

joestringer replied:
(FWIW if we print a big angry warning message to highlight that this kind of issue then this is probably workable... What I don't want to do is to hit some really weird bug where LB stops working in some vague way far in the future and find it's impossible to trace down that this was the class of issue)

brb replied:
With backend IDs of the u32 type it should be less of a problem.

Let me address this in a separate PR in which I'm making `UpdateService` idempotent.

joestringer commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -348,15 +247,55 @@ func updateMasterService(fe ServiceKey, nbackends int, nonZeroWeights uint16) er
+
+	svc, addedBackends, removedBackendIDs, err := cache.prepareUpdate(fe, backends)
+	if err != nil {
+		return err
```
Similarly here, there was acquisition above but no release.

brb replied:
Same reasoning applies as in my comment ^^.

joestringer commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -348,15 +247,55 @@ func updateMasterService(fe ServiceKey, nbackends int, nonZeroWeights uint16) er
+		}).Debug("Acquired backend ID")
+	}
+
+	cache.addBackendIDs(newBackendIDs)
```
I found this new hunk (last few dozen lines) a bit confusing vs. the adding that occurs inside `prepareUpdate()`. Is this hunk part of preparing for the update or not?

If I follow, this allocates a backend ID for each backend specified, then the logic inside `prepareUpdate()` actually adds the backend to the cache.

brb replied:
Sorry for my spaghetti :/

I've refactored this new hunk which hopefully it makes the code a bit more clear.

Anyway, the steps are following:

1. the hunk: Figure out which backends are new (in the system-wide scope).
2. the hunk: Acquire IDs for new backends.
3. `cache.addBackendIDs`: Populate the cache with the IDs.
4. `cache.prepareUpdate`: Figure out which backends need to be added/removed to/from the given service.

joestringer commented on pkg/maps/lbmap/bpfservice.go:
```diff
@@ -195,29 +243,169 @@ func (l *lbmapCache) prepareUpdate(fe ServiceKey, backends []ServiceValue) *bpfS
+			bpfSvc.slaveSlotByBackendAddrID[addrID] = pos
+		}
+	}
+	// Step 4: Add all backends that don't exist in the service v2 yet.
```
nit: Does step 4 need to iterate separately across the backends or could it be folded into step 3's loop?

brb replied:
I thought about merging into the step 3's loop, but I wanted to make it clear to the reader about the separation of v2 and legacy (at the cost of performance). Let me know which way you prefer.

joestringer replied:
It's fine as-is.

joestringer commented on pkg/maps/lbmap/bpfservice.go:
```diff
@@ -44,36 +46,49 @@ type bpfService struct {
+
+	// backendsV2 is a map of all service v2 backends indexed by the legacy IDs.
+	// TODO(brb) use list instead to preserve the ordering when svc backends change.
+	backendsV2 map[BackendAddrID]ServiceValue
```
The existing `uniqueBackends` field describes cardinality (`may be listed multiple times ... only once in uniqueBackends`), would you mind adding a similar comment for this field?

brb replied:
Sure, added.

joestringer commented on pkg/maps/lbmap/ipv4.go:
```diff
@@ -246,3 +298,172 @@ func NewRevNat4Value(ip net.IP, port uint16) *RevNat4Value {
+func NewBackend4Value(ip net.IP, port uint16, proto u8proto.U8proto) (*Backend4Value, error) {
+	ip4 := ip.To4()
+	if ip4 == nil {
+		return nil, fmt.Errorf("Not an IPv4 address")
```
Can we push this validation up closer to where we receive inputs to avoid having sanitization errors deep in the datapath reconfiguration logic?

brb replied:
Removed the redundant check in the `NewBackend{4,6}(...)` methods.

However, I'd prefer to keep the check in `NewBackend{4,6}Value(...)`, because in v1.6 its creation step will change (currently it is derived from `ServiceValue` which IP is checked). So, we could move the validation to a higher level in v1.6.

joestringer commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -384,8 +323,31 @@ func UpdateService(fe ServiceKey, backends []ServiceValue, addRevNAT bool, revNA
+			return err
+		}
+		if err := updateBackend(b); err != nil {
+			return err
```
AFAICT if we end up erroring out here, we'll just leave the LB stuff in some weird intermediate state and never attempt to configure the services again... can we improve on this? (Maybe the existing code is similar.. doesn't mean we should avoid thinking about improving this though..)

brb replied:
I was following the existing code. I could try making `UpdateService` function idempotent, so that a caller could eventually succeed to configure the service.

joestringer replied:
That'd be nice. Sounds like a future enhancement.

joestringer commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -348,15 +247,55 @@ func updateMasterService(fe ServiceKey, nbackends int, nonZeroWeights uint16) er
-func UpdateService(fe ServiceKey, backends []ServiceValue, addRevNAT bool, revNATID int) error {
+// UpdateService adds or updates the given service in the bpf maps (in both -
+// legacy and v2).
+func UpdateService(fe ServiceKey, backends []ServiceValue,
```
This function has grown massive. Any chance to split out some of the bigger chunks into shorter functional changes that can be referred to from this function? (+ ideally unit testing but maybe that's not viable)

brb replied:
I thought about this, but considering that the code got more complex and the legacy updates will go in the nearest future I didn't want to introduce more functions which IMHO would make the function less readable. I could change this though if you think otherwise.

joestringer replied:
For any function that's much longer than my screen, I struggle to keep everything in my head, and I find it more difficult to determine variable usage and potential side-effects. I won't press on this, but the longer the function is, the more incentive I have to avoid interacting with it :-).

brb replied:
OK;-)

I gave a try to make it more readable by splitting into multiple smaller functions.

joestringer commented on bpf/lib/lb.h:
```diff
@@ -656,27 +841,104 @@ static inline struct lb4_service *lb4_lookup_service(struct __sk_buff *skb,
+		struct lb4_service_v2 *svc;
+
+		/* FIXME: The verifier barks on these calls right now for some reason */
+		/* cilium_dbg_lb(skb, DBG_LB4_LOOKUP_MASTER, key->address, key->dport); */
```
FWIW I was able to place this call into the caller (`lb4_lookup_service_v2()`) and compile/verify on my own host kernel.. haven't tried this with the dev VM though.

brb replied:
The comment is a copy-paste from the `__lb{4,6}_lookup_service` functions. I've never actually tried to uncomment the lines.

joestringer commented on daemon/loadbalancer.go:
```diff
@@ -779,3 +771,109 @@ func (d *Daemon) syncLBMapsWithK8s() error {
+	}
+	svcMapV2, _, errors := lbmap.DumpServiceMapsToUserspaceV2()
+	for _, err := range errors {
+		log.WithError(err).Warning("Error occured while dumping service v2 table from datapath")
```
Won't this be guaranteed to complain the first time that Cilium is upgraded? (no such map exists)

brb replied:
Noup, it shouldn't complain, as the maps are opened or created by the daemon in `d.initMaps()` before `restoreServices()` has been called.

joestringer commented on daemon/loadbalancer.go:
```diff
@@ -779,3 +771,109 @@ func (d *Daemon) syncLBMapsWithK8s() error {
+			addRevNAT := true
+			revNATID := int(svc.FE.ID)
+			err = lbmap.UpdateService(fe, besValues, addRevNAT, revNATID,
+				service.AcquireBackendID, service.DeleteBackendID)
```
I think I see why this `service` dependency injection is necessary, though I can't help but wonder why the BPF lbmap handling code needs to deal with these details. Maybe something to think about for future cleanup.

brb replied:
Yeah, ideally the backend IDs acquisition should happen in `daemon/loadbalancer.go` to have a clear separation between layers. However, that would have required quite a bit of changes to the lbmap code which existed before my changes. Thus, I was reluctant to do any refactoring due to the PR size.

Anyway, it's in my TODO list for v1.6.

joestringer commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -622,8 +509,11 @@ func DumpServiceMapsToUserspace(includeMasterBackend bool) (loadbalancer.SVCMap,
-		svc := newSVCMap.AddFEnBE(fe, be, svcKey.GetBackend())
-		newSVCList = append(newSVCList, svc)
+		// Do not include master services
+		if svcKey.GetBackend() != 0 {
```
Is there a reason to swap this from the previous place in the previous commit? The comment a few lines up for the `RevNatKey().GetKey()` logic states that master key is not useful for constructing the cache, so I don't see why this test needs to move down to here. Can we skip the parsing/idcache population bits for masters by just simplifying the conditional statement that was removed in this commit on line 605 or so (at the start of this block)?

brb replied:
From what I reverse-engineered, it seems that the intention for the `includeMaster` flag was to to indicate that a service ID needs to be extracted when dumping the service maps. So, I didn't want skip the parsing and population bits to avoid loosing service ID during the restoration.

joestringer replied:
That's about as far as I follow... but in the `SyncLBMap()` case we ignored master services previously. Including master services there wasn't explicitly called out in a commit message which is why I brought this up.

brb replied:
Initially, I was afraid to miss a case when some old Cilium version set `revNATID` only for a master entry, but not for backend entries. However, I've just checked the commit history, and I couldn't find that it was ever actually the case. Going to change as you suggested.

joestringer commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -147,24 +65,28 @@ func updateService(key ServiceKey, value ServiceValue) error {
 	}
-	err = lookupAndDeleteServiceWeights(key)
-	if err == nil {
-		cache.delete(key)
```
This was removed from the common case, which means that the `UpdateService()` caller no longer removes the key from the cache when removing old backends that are no longer needed. Is this intentional?

brb replied:
* `cache.delete(key)` removes the service entry from `lbmapCache.entries`.
* `deleteServiceLocked(key)` is called from `UpdateService` when removing legacy service entries. This can happen when a svc legacy is updated with an empty list of endpoints before its removal (such sequence of events has been observed on k8s).

So, this can lead to situations where `lbmapCache.removeServiceV2(..)` fails due to the svc cache entry not found due to ^^. We should only delete the cache entry upon a svc removal. I'm planning to backport this fix to v1.3/v1.4.

joestringer commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -425,152 +387,85 @@ func UpdateService(fe ServiceKey, backends []ServiceValue, addRevNAT bool, revNA
+	for _, backendID := range removedBackendIDs {
+		backendKey.SetID(backendID)
+		if err := deleteBackendLocked(backendKey); err != nil {
+			return fmt.Errorf("Unable to delete backend with ID %d: %s", backendID, err)
```
Should this continue attempting to remove other backends even if one fails?

brb replied:
Yep, but it's probably part of https://github.com/cilium/cilium/pull/7540#discussion_r271785436.

joestringer reviewed (changes requested):
Nice work!

> For backend ID type I'm using u16 which sets a limit of max 65535 backends (service endpoints) per cluster. It's not too late, I can easily switch to u32 to support more backends.

If there's even a small chance that we'll want to support this, then we should probably go for it to minimize future map element resize upgrade complications. Current scale targets are 2K nodes / 20K pods which isn't that far off if most pods in the cluster are behind a service. Put a few such clusters together and you're getting pretty close.

> Currently, we maintain both types of services at the cost of larger memory footprint and more expensive service lookups. I'm adding the --disable-legacy-services to the cilium-agent to disable legacy services (will come in a separete PR). Also, in that PR I'm going to change cilium bpf lb list to use svc v2 instead of legacy as a source.

For CLI cleanliness, please name this `--enable-legacy-services`. Naming options with negative names often ends up biting us later.

Is the upgrade logic documented somewhere? I see usage of the `rx_bytes` field in service CT entries and the count field in v2 backend entries. I understand if this is intended to be part of a subsequent follow-up PR.

Most of my comments below are fairly trivial bits that could be added to a project as notes to follow up on later. I also still spot lots of FIXME's and TODO's, are you still planning to address any of these before finalising this PR?

There doesn't appear to be any meaningful new testing. I presume this is primarily because most of the functionality should be transparently replacing the previous logic. Still, given some of the new management stuff like reference counts on backends and whatnot that maybe there should be new tests to validate that these references are being properly tracked and removed to avoid leaking the backend ID address space?

The `UpdateService()` error cases seem a little odd to me from balancing allocation/deallocation.

ianvernon commented on daemon/loadbalancer.go:
```diff
@@ -246,6 +246,17 @@ func (d *Daemon) svcDelete(svc *loadbalancer.LBSVC) error {
 }
 
 func (d *Daemon) svcDeleteBPF(svc loadbalancer.L3n4AddrID) error {
+	if err := lbmap.DeleteServiceV2(svc, service.DeleteBackendID); err != nil {
```
If this fails, then the deletion of the legacy service will not occur - why not delete both, and synthesize an error in the case that only one failed, or if both failed to be deleted?

brb replied:
Sure, good suggestion. Changed.

ianvernon reviewed (approved):
Some questions / nitpicks - feel free to address these in follow-ups.

brb committed:
lbmap: Refactor UpdateService

This commit splits `UpdateService` into multiple smaller functions
for the sake of readability.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
lbmaps: Use RWMutex for lbmapCache

As we have a couple read-only methods which are quite often used,
replace `lock.Mutex` with `lock.RWMutex` to support read locks.

Signed-off-by: Martynas Pumputis <m@lambda.lt>
